### ISSUE TITLE
SCons does not remove dangling symlink target before build

### ISSUE THREAD

**Issue by roblub** [SCons/scons#3516]
Hi

According to https://scons.org/doc/3.1.2/HTML/scons-user/ch13.html, _when SCons determines that an target file needs to be rebuilt and removes the existing version of the target before executing_.

Unfortunately the above is not true, when the target file for some reason is a symbolic link to non-existing file. In such a case scons leaves the broken symlink as is, and then runs target build action, which then might try to create target file in the destination of the symlink (for example g++ works like this).

Is there any possibility to convince scons to remove the target's dangling symlink before starting  build?

**Comment by bdbaddog** [comment]
Do you have a small test case to demonstrate this issue?

**Comment by roblub** [comment]
Yes, I'm attaching a simple test case with two broken symlinks, that are not removed by scons, what causes g++ to follow them. 
[test-case.tar.gz](https://github.com/SCons/scons/files/4027818/test-case.tar.gz)

**Comment by bdbaddog** [comment]
In your test case the symlink isn't created by SCons.. 
Is that the actual situation you're working with? Or is SCons creating the symlink for you?

**Comment by roblub** [comment]
Hi,

What I am trying to do is using symlinks to files in cache instead of copying them. (I know hardlinks would be much better, but the cache is on separate storage, so they can't be used.) The aim is to have better performance and lower disk usage. 

Anyway, I've already created my own versions of  SCons.CacheDir.CacheRetrieve & CacheRetrieveSilent actions, in which I create those symlinks. I even remove dangling symlinks from there as well. 

I'm still testing this solution, but everything seems to work pretty well so far unless the --cache-disable option is used, after the link target had already been removed, for example by a cron job that clears old files from the cache. In such a case g++ happily creates new file via the symlink in the cache. That's why I am seeking for solution that would allow me to remove the broken symlinks  before g++ is run. 

I've recently learnt about AddPreAction(), but I am not sure if is it possible to add such action for any target. I do not know scons so well yet to get a list of all targets. But maybe it would be simpler if scons could somehow  remove such broken links on its own? I think it must use os.path.exists() somewhere; maybe it would be easy to use to replace it with  os.path.lexists()?

Thanks,
Robert
